feat: Fast buffer reader and fast buffer writer

ShadauxCat opened the pull request:
Covered by a total of 646 new unit tests.

ShadauxCat committed:
FastBufferWriter implemented and tested (still need to add and update some xmldoc comments though)

ShadauxCat committed:
A few additional tests to cover the last missed cases I can think of (aside from INetworkSerializable, which isn't ready to be tested yet due to lack of BufferSerializer)

ShadauxCat committed:
FastBufferReader + tests

ShadauxCat committed:
- More tests
- Streamlined the implementations of bit-packing uints and ulongs.

ShadauxCat committed:
- Removed NativeArray from FastBufferReader and FastBufferWriter, replacing with byte* allocated directly through UnsafeUtility.Malloc() (same function used under the hood by NativeArray). This is required in order to be able to store pointers to FastBufferReader and FastBufferWriter in order to be able to wrap them with BufferSerializer - NativeArray contains a managed variable in it that disallows taking a pointer to it. And since FBW and FBR are structs, pointers are the only way to wrap them "by reference" in another struct - ref fields aren't allowed even inside ref structs.

ShadauxCat committed:
Added utility ref struct "Ref<T>" to more generically support wrapping values in other ref structs in a capture-by-reference style, updated BitReader and BitWriter to use that. Also aggressively inlining properties in FBW and FBR.

ShadauxCat committed:
BufferSerializer and tests.

ShadauxCat committed:
Removed unnecessary comment.

ShadauxCat committed:
XMLDocs + cleanup for PR

ShadauxCat committed:
Merge branch 'develop' into feature/fast_buffer_reader_writer

ShadauxCat committed:
Replaced possibly unaligned memory access with UnsafeUtility.MemCpy... it's a little slower, but apparently some platforms won't support unaligned memory access (e.g., WEBGL, ARM processors) and there's no compile time way to detect ARM processors since the bytecode is not processor-dependent... the cost of the runtime detection would be more expensive than the cost of just doing the memcpy.

LukeStampfli commented on com.unity.netcode.gameobjects/Runtime/Utility/Ref.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Unity.Multiplayer.Netcode
+{
+    public struct Ref<T> where T: unmanaged
```
This entire struct is just the & operator unless I misunderstand something. There's not need for the pinning and fixing here and instead of IsSet we can just check the pointer for null?

ShadauxCat replied:
It's a compile error if I try to take the pointer directly without the fixed() expression, unfortunately. That's because even value types can be members of managed types and there's no way to add a constraint that they must be on stack memory. Ref<> is just... easier to read and avoids having to repeat the fixed() expression every time this pattern emerges.

ShadauxCat replied:
Also, Ref<> contains the unsafe code so the structs using it don't have to declare all their methods to be unsafe.

ShadauxCat committed:
Resurrected BytewiseUtil.FastCopyBytes as a faster alternative to UnsafeUtility.MemCpy for small values, while still supporting unaligned access.

ShadauxCat committed:
Reverting an accidental change.

ShadauxCat committed:
Removed files that got accidentally duplicated from before the rename.

ShadauxCat committed:
Standards fixes

ShadauxCat committed:
Removed accidentally added files.

ShadauxCat committed:
Added BuildInfo.json to the .gitignore so I stop accidentally checking it in.

JesseOlmer commented on testproject/Assets/StreamingAssets/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/BuildInfo.json
```
There's already 3+ .gitignores in the repo can we add this to one of those instead of this deeply nested one?

0xFA11 replied:
+1, could add `StreamingAssets/` line to `testproject/.gitignore` file :)

ShadauxCat replied:
Sorry about that - my git client defaults to ignoring in the same directory when selecting to ignore files. I'll make sure to edit the .gitignore manually next time instead of using the client.

0xFA11 replied:
so, let's revert this file?

ShadauxCat replied:
wtf... I did delete it... why is it still here?

JesseOlmer commented on com.unity.netcode.gameobjects/Tests/Editor/Serialization/FastBufferReaderTests.cs:
```diff
@@ -0,0 +1,2192 @@
+using System;
```
Between this file and the equivalent Writer file there's >4000 lines, and only 150 differences... Does a base class or shared utility class make sense here to get a bit more DRY?

ShadauxCat replied:
Yeah... I actually originally started by making FastBufferReaderTests be a child of FastBufferWriterTests but it wasn't really working out and I didn't want to spin my wheels on it. For some reason it didn't occur to me that making them both inherit from a shared base would work. D'oh.

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitCounter.cs:
```diff
@@ -0,0 +1,126 @@
+        {
+            0/8+1,  1/8+1, 17/8+1,  2/8+1, 18/8+1, 50/8+1,  3/8+1, 57/8+1,
+            47/8+1, 19/8+1, 22/8+1, 51/8+1, 29/8+1,  4/8+1, 33/8+1, 58/8+1,
+            15/8+1, 48/8+1, 20/8+1, 27/8+1, 25/8+1, 23/8+1, 52/8+1, 41/8+1,
+            54/8+1, 30/8+1, 38/8+1,  5/8+1, 43/8+1, 34/8+1, 59/8+1,  8/8+1,
+            63/8+1, 16/8+1, 49/8+1, 56/8+1, 46/8+1, 21/8+1, 28/8+1, 32/8+1,
+            14/8+1, 26/8+1, 24/8+1, 40/8+1, 53/8+1, 37/8+1, 42/8+1,  7/8+1,
+            62/8+1, 55/8+1, 45/8+1, 31/8+1, 13/8+1, 39/8+1, 36/8+1,  6/8+1,
+            61/8+1, 44/8+1, 12/8+1, 35/8+1, 60/8+1, 11/8+1, 10/8+1,  9/8+1,
+        };
+
+        private static readonly int[] k_DeBruijnTableBytes32 =
+        {
+            0/8+1,  1/8+1, 16/8+1,  2/8+1, 29/8+1, 17/8+1,  3/8+1, 22/8+1,
+            30/8+1, 20/8+1, 18/8+1, 11/8+1, 13/8+1,  4/8+1,  7/8+1, 23/8+1,
+            31/8+1, 15/8+1, 28/8+1, 21/8+1, 19/8+1, 10/8+1, 12/8+1,  6/8+1,
+            14/8+1, 27/8+1,  9/8+1,  5/8+1, 26/8+1,  8/8+1, 25/8+1, 24/8+1,
+        };
+
+        // And here we're counting the number of set bits, not the position of the highest set,
+        // so these still have +1 applied - unfortunately 0 and 1 both return the same value.
+        private static readonly int[] k_DeBruijnTableBits64 =
+        {
+            0+1,  1+1, 17+1,  2+1, 18+1, 50+1,  3+1, 57+1,
+            47+1, 19+1, 22+1, 51+1, 29+1,  4+1, 33+1, 58+1,
+            15+1, 48+1, 20+1, 27+1, 25+1, 23+1, 52+1, 41+1,
+            54+1, 30+1, 38+1,  5+1, 43+1, 34+1, 59+1,  8+1,
+            63+1, 16+1, 49+1, 56+1, 46+1, 21+1, 28+1, 32+1,
+            14+1, 26+1, 24+1, 40+1, 53+1, 37+1, 42+1,  7+1,
+            62+1, 55+1, 45+1, 31+1, 13+1, 39+1, 36+1,  6+1,
+            61+1, 44+1, 12+1, 35+1, 60+1, 11+1, 10+1,  9+1,
+        };
+
+        private static readonly int[] k_DeBruijnTableBits32 =
+        {
+            0+1,  1+1, 16+1,  2+1, 29+1, 17+1,  3+1, 22+1,
+            30+1, 20+1, 18+1, 11+1, 13+1,  4+1,  7+1, 23+1,
+            31+1, 15+1, 28+1, 21+1, 19+1, 10+1, 12+1,  6+1,
+            14+1, 27+1,  9+1,  5+1, 26+1,  8+1, 25+1, 24+1,
+        };
```
Is the repeated +1 better than looping through and adding 1 in a static constructor? (Honest question, not making a suggestion - LUTs are necessarily rigid anyways)

ShadauxCat replied:
I think it is because you can forget to call the static initializer... you can't forget to have a constant. :) Plus, even for something that only happens once, compile time math beats runtime math.

becksebenius-unity commented on testproject/Assets/Tests/Runtime/NetworkSceneManagerTests.cs:
```diff
@@ -11,7 +11,7 @@
 namespace TestProject.RuntimeTests
 {
-    public class NetworkSceneManagerTests: BaseMultiInstanceTest
+    public class NetworkSceneManagerTests : BaseMultiInstanceTest
```
Just nitpicking, this file doesn't have any actual changes, so for a big review like this I think it would be in your best interest to omit these types of changes

ShadauxCat replied:
dotnet-format did this :/... it changed a lot of files when I ran standards.py --fix and I didn't look through to see whether all of them were mine before I checked in... Sorry about that... but at the same time... probably worth getting develop to a point where standards.py --fix does nothing to it? Maybe in a different PR though...

becksebenius-unity replied:
Yeah, my point is mainly about keeping PRs focused, this file is a distraction and makes the PR look bigger than it actually is, so it would help with review traction to do a quick audit for these kinds of things before putting it up

0xFA11 replied:
we sometimes get to that "`./standards.py --fix` does nothing" point but people often do not run `./standards.py --fix` locally before they push their changes (even though you could do `./standards.py --hook` for it) and we get these whitespace updates whenever somebody else does and fixes things.

becksebenius-unity commented on com.unity.netcode.gameobjects/Tests/Editor/Serialization/BitWriterTests.cs:
```diff
@@ -0,0 +1,333 @@
+
+                Assert.AreEqual(0, *asInt);
+
+                Assert.Throws<OverflowException>(() =>
+                {
+                    using (var bitWriter = writer.EnterBitwiseContext())
+                    {
+                        bitWriter.WriteBit(true);
+                    }
+                });
+
```
Why are these all in the same test?

ShadauxCat replied:
The test is "verify trying to write without bounds checking throws an exception"... it's checking that for all the functions that should throw an exception. I would argue that this isn't four different tests of writing bits, it's one test of VerifyCanWriteBits

becksebenius-unity replied:
I see, I think this is a difference in test strategy and a discussion for another time - I'm a proponent of one test per case, even though they're similar tests, they're different cases.

ShadauxCat replied:
I started breaking things apart as per your suggestion but it's taking time. I'm going to stop this effort for now so I can start on IMessage given our deadlines at the moment. I'm making a Jira to come back and continue addressing these.

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/IBufferSerializerImplementation.cs:
```diff
@@ -0,0 +1,39 @@
+{
+    public interface IBufferSerializerImplementation
+    {
+        public bool IsReader { get; }
+        public bool IsWriter { get; }
+
+        public ref FastBufferReader GetFastBufferReader();
+        public ref FastBufferWriter GetFastBufferWriter();
+
+        public void SerializeValue(ref object value, Type type, bool isNullable = false);
+        public void SerializeValue(ref INetworkSerializable value);
+        public void SerializeValue(ref GameObject value);
+        public void SerializeValue(ref NetworkObject value);
+        public void SerializeValue(ref NetworkBehaviour value);
+        public void SerializeValue(ref string s, bool oneByteChars = false);
+        public void SerializeValue<T>(ref T[] array) where T : unmanaged;
+        public void SerializeValue(ref byte value);
+        public void SerializeValue<T>(ref T value) where T : unmanaged;
+
+        // Has to have a different name to avoid conflicting with "where T: unmananged"
+        // Using SerializeValue(INetworkSerializable) will result in boxing on struct INetworkSerializables
+        // So this is provided as an alternative to avoid boxing allocations.
+        public void SerializeNetworkSerializable<T>(ref T value) where T : INetworkSerializable;
+
+        public bool PreCheck(int amount);
+        public void SerializeValuePreChecked(ref GameObject value);
+        public void SerializeValuePreChecked(ref NetworkObject value);
+        public void SerializeValuePreChecked(ref NetworkBehaviour value);
+        public void SerializeValuePreChecked(ref string s, bool oneByteChars = false);
+        public void SerializeValuePreChecked<T>(ref T[] array) where T : unmanaged;
+        public void SerializeValuePreChecked(ref byte value);
+        public void SerializeValuePreChecked<T>(ref T value) where T : unmanaged;
```
Nit (sorry :D ) but `public` is extraneous for interfaces

ShadauxCat replied:
yeah... I keep forgetting that... please pardon me, I've been coding in c# for exactly three months now. 😅

becksebenius-unity replied:
no worries. while you're at it - there's a lot of `private` syntax but our coding standards say those should be omitted. (I find this one hard myself.... although I'm surprised that your IDE isn't screaming at you about these if our editor config is set up?)

ShadauxCat replied:
Rider's editorconfig plugin doesn't work correctly, unfortunately.

ShadauxCat replied:
The `private` is throughout the whole codebase. I don't think as a team we're following a standard of not using `private` - I'd value consistency over correctness in this case, so I'm not going to change that in my PR. If we want to change that, we should make a pass over the whole codebase as a separate task.

NoelStephensUnity replied:
@MFatihMAR Our standards checks require we specify private correct?

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitReader.cs:
```diff
@@ -0,0 +1,218 @@
+        /// <summary>
+        /// Pads the read bit count to byte alignment and commits the read back to the reader
+        /// </summary>
+        public void Dispose()
```
Did you mean to implement IDisposable?

ShadauxCat replied:
ref structs can't implement interfaces, but c# has special behavior to support `using()` on ref structs that have a `public void Dispose()`

becksebenius-unity replied:
Ohh, neat. ref structs are pretty new to me. I'm a little uneasy about some of your changes in general because I haven't dipped into some of the new ref syntax - especially the Ref<T> type you're adding. I hope you're seeking review from some folks that have a deep experience with this since this type of code is very risky

becksebenius-unity replied:
(To be clear, not saying you shouldn't be doing it, serialization is exactly where I'd expect to be writing this type of code)

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitWriter.cs:
```diff
@@ -0,0 +1,206 @@
+        /// This exists as a separate method to allow multiple bit writes to be bounds checked with a single call.
+        /// If it returns false, you may not write, and in editor and development builds, attempting to do so will
+        /// throw an exception. In release builds, attempting to do so will write to random memory addresses and cause
+        /// Bad Things(TM).
```
nit but since these API comments go online as a learning resource I'd suggest a more professional "and cause unexpected behavior".

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitWriter.cs:
```diff
@@ -0,0 +1,206 @@
+                }
+            }
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            m_AllowedBitwiseWriteMark = newBitPosition;
```
I find it a little odd that a `Verify` method (which sounds like a pure function) mutates state. Nothing in the XML comment suggests that state is modified either.

ShadauxCat replied:
The comment says:

```c#
        /// Unsafe write operations will throw OverflowException in editor and development builds if you
        /// go past the point you've marked using VerifyCanWrite(). In release builds, OverflowException will not be thrown
        /// for performance reasons, since the point of using VerifyCanWrite is to avoid bounds checking in the following
        /// operations in release builds.
```

Does that not indicate that VerifyCanWrite is going to mutate state in editor and development builds? In release builds, it doesn't... it only does it here as a safeguard to alert developers when they forgot to do bounds checking.

becksebenius-unity replied:
Where is that text coming from?

becksebenius-unity replied:
Also, I started this comment here but I see above that state is mutated in non-development contexts too (`m_BufferPointer = m_Writer.Value.BufferPointer + m_Writer.Value.PositionInternal;`)

ShadauxCat replied:
Oh, sorry, wrong Verify method... that was in FastbufferWriter. This one just says:

```c#
        /// If it returns false, you may not write, and in editor and development builds, attempting to do so will
        /// throw an exception.
```

I'll update this text to match the main reader/writers.

becksebenius-unity replied:
Sounds good. I'd suggest workshopping the Verify name though. At face value, it screams const to me. I don't have any specific suggestions though and I'll leave it to your discretion.

ShadauxCat replied:
> Also, I started this comment here but I see above that state is mutated in non-development contexts too > >(m_BufferPointer = m_Writer.Value.BufferPointer + m_Writer.Value.PositionInternal;)

Right, also forgot to mention in the comment on this one that the buffer will grow if there's not enough space. That's what that part is. Instead of returning false and making you grow it yourself, it grows it for you to accommodate your request and then returns true.

That's done for performance reasons so you can do one big bounds check (and grow if necessary) before writing 20 fields, instead of writing 20 fields with 20 bounds checks slowing them all down.

ShadauxCat replied:
Perhaps `BeginWrite()` is a better name? But there's no corresponding `EndWrite()` so that might be confusing...

becksebenius-unity replied:
I do like that, it informs that this should be called in sequence also it is pretty obvious that it will modify some state. Although it doesn't really describe the return value

becksebenius-unity replied:
TryBeginWrite? :P

JesseOlmer replied:
Plenty of APIs have nop EndWrites. You could also do something like ReserveBytes which gets across the mutative behavior here, but not the required-order aspect.

ShadauxCat replied:
I think TryBeginWrite is the most clear. I'll switch it to that.

JesseOlmer commented on com.unity.netcode.gameobjects/Tests/Editor/Serialization/FastBufferWriterTests.cs:
```diff
@@ -0,0 +1,2191 @@
+            if (testType == typeof(byte))
+            {
+                byte b = (byte)random.Next();
+                if (writeType == WriteType.WriteDirect)
```
nit: You can probably save yoursel some 2-300 lines of repeats here by using a local function like

```
            void RunTypeTestLocal<T>(T val, WriteType wt) where T : unmanaged
            {
                switch (wt)
                {
                    case WriteType.WriteDirect:
                        RunTypeTest(val);
                        break;
                    case WriteType.WriteSafe:
                        RunTypeTestSafe(val);
                        break;
                    default:
                        RunObjectTypeTest(val);
                        break;
                }
            }
```

ShadauxCat replied:
This is a very good idea. Figuring out ways to be more DRY is definitely my weak point as a programmer. Thank you for the suggestions.

Cosmin-B commented on com.unity.netcode.gameobjects/Runtime/Serialization/BytePacker.cs:
```diff
@@ -0,0 +1,618 @@
+            }
+            if (value is GameObject)
+            {
+                var networkObject = ((GameObject)value).GetComponent<NetworkObject>();
```
Hey @ShadauxCat I suggest using TryGetComponent instead of GetComponent as the former does not allocate memory (whilst in Editor)

ShadauxCat replied:
Wow. I didn't know GetComponent() allocated. Thanks for the heads up.

NoelStephensUnity replied:
I just learned that too.  Nice Cosmin!

Cosmin-B commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -0,0 +1,1042 @@
+        /// <param name="value">The value to write</param>
+        public void WriteValue(GameObject value)
+        {
+            var networkObject = (value).GetComponent<NetworkObject>();
```
Hey @ShadauxCat I suggest converting this GetComponent call to TryGetComponent, as the latter does not allocate anything (in Editor)

Cosmin-B commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -0,0 +1,1042 @@
+        /// <param name="value">The value to write</param>
+        public void WriteValueSafe(GameObject value)
+        {
+            var networkObject = (value).GetComponent<NetworkObject>();
```
Hey @ShadauxCat I suggest converting this GetComponent call to TryGetComponent, as the latter does not allocate anything (in Editor)

LukeStampfli commented on com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs:
```diff
@@ -0,0 +1,313 @@
+    ///
+    /// BufferSerializer doesn't wrapp FastBufferReader or FastBufferWriter directly because it can't.
+    /// ref structs can't implement interfaces, and in order to be able to have two different implementations with
+    /// the same interface (which allows us to avoid an "if(IsReader)" on every call), the thing directly wrapping
```
I'm trying to understand how we would end up using the BufferSerializer and IBufferSerializerImplementation. From what I can tell currently they are unused.

So if we were to replace `NetworkSerializer` with `BufferSerializer` we would have to pass the interface to the `NetworkSerialize` function of `INetworkSerializable`. This will incur boxing which is much worse than a `if(IsReader)` check which most of the time will be optimized away by branch prediction anyways.

ShadauxCat replied:
Yes, everything in this PR is currently unused. 

The plan is for INetworkSerializable's interface to change to be like this: 

```c#
interface INetworkSerializable
{
    void Serialize<T>(BufferSerializer<T>) where T: IBufferSerializerImplemetation;
}
```
BufferSerializer is a ref struct - c# language rules say it can NEVER be boxed. The generic function signature not only prevents boxing, but is actually the only way you can ever receive this type. And since `m_Implementation` is of a generic type rather than of IBufferSerializerImplementation, it also won't be boxed.

LukeStampfli replied:
The above code you sent won't work right? what is needed would be:

```cs
internal interface INetworkSerializable<T> where T : IBufferSerializerImplementation
{
    void Serialize(BufferSerializer<T> serializer) ;
}
```

And then as a user I would need to implement this generic interface for all instances so essentially:

```cs
 internal struct Foo: INetworkSerializable<BufferSerializerReader>, INetworkSerializable<BufferSerializerWriter>
  {
      public int Bar;

      public void Serialize(BufferSerializer<BufferSerializerReader> serializer)
      {
          serializer.Serialize(ref Bar);
      }

      public void Serialize(BufferSerializer<BufferSerializerWriter> serializer)
      {
          serializer.Serialize(ref Bar);
      }
  }
```

And sure this works but it's ugly and defeats the point of having a serializer interface instead of a reader/writer interface.

becksebenius-unity replied:
+1, we should continue to have a ReaderWrite interface if at all possible because it makes serialization code symmetrical by default, which is a huge win for users writing custom serializers

ShadauxCat replied:
No, the code I sent does work:

```c#
interface IBufferSerializable
{
    void Serialize<T>(BufferSerializer<T> serializer) where T : IBufferSerializerImplementation;
}

class Foo : IBufferSerializable
{
    public void Serialize<T>(BufferSerializer<T> serializer) where T : IBufferSerializerImplementation
    {
        bool b = true;
        serializer.SerializeValue(ref b);
    }
}

static void SerializeIBufferSerializable(IBufferSerializable b)
{
    FastBufferWriter writer = new FastBufferWriter(100, Allocator.Temp);
    BufferSerializer<BufferSerializerWriter> serializer =
        new BufferSerializer<BufferSerializerWriter>(new BufferSerializerWriter(ref writer));
    b.Serialize(serializer);
}
```

It's still a read/write interface. It's still symmetrical. I just wrote that code in rider and it compiled fine, and no boxing.

ShadauxCat replied:
I promise it works. I tested it before I submitted the PR. I didn't keep that particular test code because I wasn't ready to change the definition of INetworkSerializable yet, but it does work.

LukeStampfli replied:
Ah I see, you are totally right, somehow missed that you can just keep the function as generic in the interface. Ok I'm completely fine with this.

LukeStampfli commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReader.cs:
```diff
@@ -0,0 +1,987 @@
+        {
+            ReadValueSafe(out ulong networkObjectId);
+
+            if (NetworkManager.Singleton.SpawnManager.SpawnedObjects.TryGetValue(networkObjectId, out NetworkObject networkObject))
```
This singleton use here will cause issues with multi instance test where this will always reference the same instance of a NetworkObject instead of the correct one.

LukeStampfli replied:
Same goes for NetworkObject, NetworkBehaviour read/writes. Could we consider scrapping them until we have a need for them and then implement a better solution?

ShadauxCat replied:
yep - I'm aware of that. I put in the singleton for now because it's not hooked up to anything yet and I wanted to be able to test the basic functionality more easily. I was going to fix that once I get further into the IMessage implementation and see how everything is hooking together.

Though, honestly, I'm not sure I like having NetworkObject and NetworkBehaviour in the readers/writers at all. There's been some requests to make these available generically in their own package some day... these functions would make that impossible. But I don't have a good alternative to suggest that wouldn't feel odd and clunky.

Actually, putting these things in extension methods seems like a good solution. Also a good way for users to add `WriteValue()` support for their own types... should make a note to document that as an option.

LukeStampfli commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReader.cs:
```diff
@@ -0,0 +1,987 @@
+        /// for multiple reads at once by calling VerifyCanRead.
+        /// </summary>
+        /// <param name="array">Stores the read array</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Just a note, the JITter won't inline this because the function contains throw statements.

ShadauxCat replied:
Really... that's interesting.

ShadauxCat replied:
Do you have a link to documentation on this? I can't find anything about this from my weak google-fu. I'd be interested in reading about what sorts of things prevent inlining in C#.

LukeStampfli replied:
There is no coherent official documentation for this AFAIK. It also depends on the JIT. Mono JIT (Unity) and Ryu JIT inline in a very different way and I have no idea what IL2CPP does but I hope it considers the `AggressiveInlining`.

This article is pretty accurate about how inlining in c# works: https://devblogs.microsoft.com/premier-developer/a-common-execution-path-optimization/

ShadauxCat replied:
This is good to know, and the article points out I can move the exception throwing part to a separate method to get around this. I'll make a Jira to come back and address that to make sure everything in here can be inlined.

andrews-unity replied:
IL2CPP does honor it

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitWriter.cs:
```diff
@@ -0,0 +1,206 @@
+        /// </summary>
+        /// <param name="value">Value to get bits from.</param>
+        /// <param name="bitCount">Amount of bits to write</param>
+        public unsafe void WriteBits(ulong value, int bitCount)
```
uint bitcount gets rid of negative bounds checking?

ShadauxCat replied:
Good call.

LukeStampfli reviewed (approved):
Everything I mentioned has been addressed. This is great stuff can't wait to replace the existing reader,writer and serializers with this.

I learned a lot about special use cases for ref structs and generic types while reading this PR. And thanks @ShadauxCat for keeping up with all my incorrect remarks 👍

ShadauxCat committed:
Addressed most of the review feedback. Still need to do a little more restructuring of some of the other tests.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Utility/Ref.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Unity.Multiplayer.Netcode
+{
+    public struct Ref<T> where T : unmanaged
```
I have mixed feelings about this and `RefArray` :/

ShadauxCat replied:
RefArray isn't used yet, but it has a different purpose. It will allow a stackalloc'd array to be passed to a function accepting IReadOnlyList - should be even better than NativeArray for passing client IDs to SendMessage(). 

Technically I suppose RefArray could have been left out of this PR since it's not used yet.

ShadauxCat replied:
As far as Ref goes... It's just syntactic sugar around storing a pointer in a struct. I thought it made the things using it a little easier to read. But I'm not gonna die on the hill of Ref if it's not liked.

ShadauxCat replied:
Just found another use for Ref: A dictionary of value types can hold the types by reference if it's `Dictionary<ulong, Ref<T>>` but you can't do `Dictionary<ulong, T*>` - it's a compile error.

andrews-unity replied:
But you could just use NativeHashMap or UnsafeHashMap, which do support exactly this.

ShadauxCat replied:
Oh. I didn't realize those were a thing. That would be better, yes.

ShadauxCat replied:
...ok, how do I do this with NativeHashMap or UnsafeHashMap? `T*` is still not a valid template parameter at the C# language level. I thought you were saying they'd let me access elements by ref, but I see no API in either that would let me do so.

ShadauxCat replied:
This conversation isn't really relevant to the changes here so I'm going to resolve it. `RefArray` has already been removed.

ShadauxCat committed:
standards.py --fix

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BytewiseUtility.cs:
```diff
@@ -0,0 +1,80 @@
+    {
+        /// <summary>
+        /// Helper function optimized for quickly copying small numbers of bytes.
+        /// Faster than UnsafeUtil.Memcpy and other alternatives for amount <= 8
```
You mentioned perf issues here, did you test this in the editor or the player? I ask because we have a lot of checks in this code gated by UNITY_COLLECTION_CHECKS which is always on in the editor but is never on in the player in release non-development builds. 

The overhead from C# to C++ marshaling would be very minimal in this case so I am not sure the perf impact is actually as bad as you may think

ShadauxCat replied:
Tested in editor, player built with mono, and player built with ilcpp. This was faster in all three cases.

ShadauxCat replied:
Timings based on doing things 10,000,000 times

In mono player:

1 byte FastCopyBytes: 3,047,773 ticks
1 byte MemCpy: 6,975,908 ticks
8 byte FastCopyBytes: 4,062,733 ticks
8 byte MemCpy: 6,713,209 ticks

In IL2CPP:

1 byte FastCopyBytes: 1,943,395 ticks
1 byte MemCpy: 4,288,755 ticks
8 byte FastCopyBytes: 2,621,003 ticks
8 byte MemCpy: 4,197,216 ticks

ShadauxCat replied:
For clarity, I'm not saying MemCpy has performance problems. This is a micro-optimization.

andrews-unity replied:
> For clarity, I'm not saying MemCpy has performance problems. This is a micro-optimization.

Code we write; is code we support/own and I am just wondering if the value of having to maintain that code is worth the marginal gains over memcpy; in that memcpy is a core Unity function and its not going anywhere its very well tested and is always going to be supported by a team that has the resources to support it and its far less complex code then what is here in that its literally a std::memcpy call. 

I guess what I am getting at is... the value of the micro-optimization is now code we have to support and own and is that worth it ? when we have something that works is a core engine function and is supported by a team that can support it and is less that we have to care about? Like reducing complexity and what we are responsible for and still getting value seems like a win to me.

ShadauxCat replied:
In a lot of cases I'd agree with you... serialization is one of the few cases where I think micro-optimizations like this are worthwhile. I don't see the maintenance cost of this being... really anything at all. It's just a series of assignment operations, and it supports exactly 8 cases, and beyond those 8 cases it falls back to MemCpy. But serialization is going to be something that's done over and over and over and over again, and this is 2x faster than MemCpy, so... I do think it's worth it, personally.

ShadauxCat replied:
(Although interestingly enough, in il2cpp builds, a simple `for` loop copying the whole amount one byte at a time is faster than both. But it's about the same as MemCpy in mono builds, so I opted to do this one as a compromise between mono performance and il2cpp performance.)

0xFA11 replied:
I'd also lean towards using Unity's MemCpy instead of crafting one for the reasons Andrew outlined above. Maybe just to add a little bit on that, it's more costly to modify/change/maintain home brew solutions in the mid/long term and it always becomes a "should we throw away this thing that we invested X amount of time & effort" compared to just sticking with what's already out there in Unity core APIs.

ShadauxCat replied:
Removed.

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs:
```diff
@@ -0,0 +1,313 @@
+        /// If true, will truncate each char to one byte.
+        /// This is slower than two-byte chars, but uses less bandwidth.
+        /// </param>
+        public void SerializeValue(ref string s, bool oneByteChars = false)
```
I wonder if we should just impose limits via FixedString and just say you can't send a large string but a string of fixed size: 

https://docs.unity3d.com/Packages/com.unity.collections@0.5/api/Unity.Collections.FixedString128.html

Cosmin-B replied:
This is an idea that I like

ShadauxCat replied:
I've been thinking about how to support reusing existing fixed-size arrays and even existing strings (which could have their data overwritten in an unsafe code block) but the issue I've been grappling with on it is, it's not always reasonable to say "there will always be exactly N elements". It's more reasonable to say "there will be _at most_ N elements" and preallocate for that, but the issue is... when you read the elements, how do you communicate to the user how many elements are actually there? Even FixedString seems to have this issue since there's no way to set a `Length` separate from a `Capacity`.

It could be something like `public void SerializeValue(ref string s, ref int length, bool oneByteChars = false)` and the equivalent for arrays, but I'm worried this will lead to confusion around people trying to use `s.Length` rather than the returned `length` value. On the other hand, I'm also concerned about the fact that reading strings and arrays causes an allocation.

Another possibility is that we don't support c# arrays and strings and we make our own `Array<T>` and `String` classes, which wrap the underlying structures and differentiate between `Length` and `Capacity`, so the reader can overwrite the data and then set the length and everything works intuitively... except that users have to know to use our classes because the C# ones aren't supported.

Open for input on the tradeoffs of those three approaches for sure.

ShadauxCat replied:
One benefit of our own array and string classes would be that we could still make them dynamic. If they're large enough to fit the data they can be reused, if the data exceeds the size they could reallocate only as needed to grow. That feels like the best solution to me, we just have to make sure the documentation is clear. Thoughts?

ShadauxCat replied:
I've been adding some stuff that fits this need in the INetworkMessage branch because I ran into the need for it there, so I'm going to resolve this and it'll be addressed there.

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs:
```diff
@@ -0,0 +1,313 @@
+        /// Serialize a NetworkObject
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValuePreChecked(ref NetworkObject value)
```
What are these PreChecked functions? its not clear from the function documentation what its suppose to do. Do they check if you have space required available

ShadauxCat replied:
See the documentation of PreCheck above... I'll update these comments to also include that information.

ShadauxCat replied:
tl;dr: FastBufferWriter/Reader have functions that require manual bounds checking ahead of time so the bounds checking can be done in a more performant way, and offer Safe versions you have to go a bit further out of your way to ask for if you don't want to do that. BufferSerializer just reverses the paradigm, safe by default but offering faster alternatives if you know what you're doing.

ShadauxCat replied:
Comments updated.

ShadauxCat committed:
standards.py --fix

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/Utility/RefArray.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace Unity.Multiplayer.Netcode
+{
+    public ref struct RefArray<T> where T : unmanaged
```
I am curious what the thinking is behind this is, I mean here again we have collections that can hold a pointer to an unmanged type we even have Unsafe ones that do not do any of the extra checking

ShadauxCat replied:
This isn't used yet but the idea is to get around Unity not supporting `Span<T>` by doing this:

```c#
unsafe
{
    ulong* clientIds = stackalloc ulong[requiredSize];
    // populate client IDs however you need to
    NetworkManager.SendMessage(message, new RefArray<ulong>(clientIds, requiredSize));
}
```

It allows us to pass a variable created via stackalloc into a function that takes `IReadOnlyList` as a parameter - which means we can pass `ulong[]` if we already have one, or we can work on the stack with no allocations if we don't.

We could also use `NativeArray<ulong>` but it doesn't support the `IReadOnlyList` interface (though we could easily make a wrapper that does). I haven't tested this yet, and definitely intend to do so before using `RefArray` in this context, but I have a strong suspicion based on the stuff I've done with NativeArray so far that creating a NativeArray and using NativeArray's `[]` operator will be slower than RefArray around a stackalloc'd array.

andrews-unity replied:
We do actually support readonly NativeArrays :D you can just: 

```
var x = new NativeArray<byte>(1, Allocator.Temp);
var ro = x.AsReadOnly(); // this returns a readonly view into the array
```

andrews-unity replied:
Also we have a Span<T> type for native collections its called NativeSlice<T> and it can do a lot the same things Span can.

Cosmin-B replied:
Hey all! Could someone help me understand what is the actual difference between these containers/wrappers and the Native ones? I am trying better understand the thinking behind why we felt like we needed them? Personally - and as Andrew mentioned - it feels like these use cases would be already covered by Native Collections.

ShadauxCat replied:
Being read-only isn't the point... The point is passing something created with stackalloc without copying it. NativeArray would create a new temp allocation and copy, which would defeat the purpose of the original stackalloc. It also doesn't address that NativeArray is slower than RefArray due to all its safeguards (and still slower even in release builds), and that NativeArray doesn't implement IReadOnlyList, which I will need.

I can fix the second with a wrapper class, but that'll likely make the first worse. But maybe I'm over-optimizing here? RefArray is a simple enough struct that I think it's worthwhile. 

Also NativeSlice doesn't replace Span because it requires a NativeArray to construct it. The reason I mentioned Span is that in mainline C# the compiler allows this: 

```c#
Span<int> foo = stackalloc int[5]();
```

Unity has no type that can be directly assigned to from a stackalloc to be able to pass the stackalloc'd memory to another function. That's the use case for RefArray - neither NativeArray nor NativeSlice enable that usage.

ShadauxCat replied:
Oh, I see why you mentioned being read-only. I don't need IReadOnlyList to make it read only. I need IReadOnlyList because it's the interface that provides `array.Count` and `array[idx]` for generic functions. NativeArray has both of those, but doesn't list IReadOnlyList among its implemented interfaces, so I can't use it generically. I don't actually care if it's read only, I care if it's able to be used in a generic function, and NativeArray is not. It does let me write a generic function that uses the `foreach` syntax, but that causes boxing allocations converting the returned enumerator to IEnumerator when used in a generic context.

ShadauxCat replied:
Tl;dr for Cosmin... The difference is, RefArray

- Can wrap an existing pointer (such as a return from stackalloc) __without any copies__
- Has less overhead since it's not intended for thread safety in any way and its functions are all aggressively inlined
- Can be used generically without incurring boxing allocations

None of those are true for NativeArray. NativeArray has a lot of good uses, but doesn't fit the specific need I'm trying to address with RefArray.

ShadauxCat replied:
To put it another way, RefArray exists to give the IReadOnlyList interface to a raw pointer and pass that pointer to a generic function. Very specific use case. Not intended to replace NativeArray for any other usage.

andrews-unity replied:
> Can wrap an existing pointer (such as a return from stackalloc) without any copies

You can do: 

```
int* foo = stackalloc int[5]();
var xx = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<int>(foo, 5, Allocator.Temp);
```
This doesn't do a copy at all.

ShadauxCat replied:
Ok. That would work for this use case - in my defense, though, that function's hard to find. :P

I still contend that RefArray will be faster than NativeArray, but maybe not enough for us to care in this case. 🤷 And I'll still have to write a wrapper for NativeArray to add `IReadOnlyList` support, so the difference would end up being

```c#
int* foo = stackalloc int[5]();
var xx = new RefArray(foo, 5);
```

vs

```c#
int* foo = stackalloc int[5]();
var xx = new NativeArrayReadOnlyListWrapper(NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<int>(foo, 5, Allocator.None));
```

The former is still a lot more readable to me, in addition to the performance gains.

But I'll remove it if you still want me to after having heard my reasoning.

0xFA11 replied:
why do we care about `IReadOnlyList` interface this much again? 👀

ShadauxCat replied:
```c#
public void SendMessage<T, U>(in T message, in U clientIds)
    where T: IMessage
    where U: IReadOnlyList<ulong>
```

That function signature allows passing whatever you have readily available. If you have `ulong[]` or `List<ulong>` already, you can pass it as is. If you need to create something, you can `stackalloc` and pass a wrapper with no allocations. Taking `ulong[]` forces allocating any time the list we need doesn't already happen to exist. 

`IReadOnlyList` is important because it supplies `array.Count` and `array[index]`. NativeArray implements IEnumerable. If that function took IEnumerable instead of IReadOnlyList, you could still pass all of those types, but every time you call `foreach` on it, you would get a boxing allocation because it has to convert the returned value to IEnumerable. Any polymorphism that isn't on a generic type parameter with a where clause will always result in boxing. Can't be zero-alloc if we box every time we send.

ShadauxCat replied:
We can get around that boxing with IReadOnlyList by using a c-style for loop and using index access instead of using foreach.

ShadauxCat replied:
After talking with Jesse, I think we've come to a good compromise on this. I'm deleting this file for now. A more focused implementation to enable the same concept will show up in the IMessage PR in a way that I think will be more palatable.

ShadauxCat committed:
Fixed incorrect namespaces.

ShadauxCat committed:
-Fixed a couple of issues where growing a FastBufferWriter wouldn't work correctly (requesting beyond MaxCapacity and requesting more than double current capacity)
-Added support for FastBufferReader to be used in a mode that doesn't copy the input buffer

ShadauxCat committed:
Fix a test failure and better implementation of large growths

NoelStephensUnity reviewed (approved):
Can't wait to see the other part that integrates all of this into the SDK!
Looks good so far!

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitReader.cs:
```diff
@@ -0,0 +1,217 @@
+            int checkPos = (int)(m_BitPosition + bitCount);
+            if (checkPos > m_AllowedBitwiseReadMark)
+            {
+                throw new OverflowException("Attempted to read without first calling TryBeginReadBits()");
```
a minor suggestion, use `nameof()` in conjunction with with string-interpolation:
```suggestion
                throw new OverflowException($"Attempted to read without first calling {nameof(TryBeginReadBits)}()");
```

ShadauxCat replied:
Done

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitReader.cs:
```diff
@@ -0,0 +1,217 @@
+            return (byte)result;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
I'm not entirely sure how much we gain from aggressive inlining attributes like these — do you have any metrics or comparison?

ShadauxCat replied:
It wasn't earth-shattering but it was significant. Like 20-30% gain IIRC. But I'll rerun those tests tomorrow.

0xFA11 replied:
> 20-30%

:O

ShadauxCat replied:
Here's the tests. Tested both aggressive inlining and branching.

Editor:
```
Inlined/No Branch: 18,707,485 ticks
Not Inlined/No Branch: 19,292,008 ticks (+3.12%)
Inlined/With Branch: 24,172,988 ticks (+29.21%)
Not Inlined/With Branch: 24,214,432 ticks (+29.43%)
```
Player:
```
Inlined/No Branch: 3,940,828 ticks
Not Inlined/No Branch: 5,748,679 ticks (+45.87%)
Inlined/With Branch: 4,384,911 ticks (+11.2%)
Not Inlined/With Branch: 7,607,113 ticks (+93.03%)
```
IL2CPP:
```
Inlined/No Branch: 2,144,231 ticks
Not Inlined/No Branch: 3,316,924 ticks (+54.69%)
Inlined/With Branch: 3,770,607 ticks (+75.84%)
Not Inlined/With Branch: 3,719,728 ticks (73.47%)
```

I think player and il2cpp numbers matter more than editor, and I'd call them significant.

andrews-unity replied:
Just for the same of completeness can we add the code used to get these perf numbers into the test project ? Might just be a good reference.

ShadauxCat replied:
Where and how should this kind of code be added to the test project, given that it's not really an automated test?

andrews-unity replied:
We have code in the test project that are manual tests... I mean its a basic unity project we have all sort of things in there its not a project restricted to just automated testing. We do also use it as the project for that when running package tests because it just makes the package ci process easier.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/BitWriter.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <summary>
+        /// Pads the written bit count to byte alignment and commits the write back to the writer
+        /// </summary>
+        public void Dispose()
```
well, `Dispose()` is often used to implement `IDisposable` interface in C# and it communicates something different than "pad bits", it's "free up resources" (which is often called from destructors when needed).
Should we rename this as `PadBits()` which would be similar to it's former predecessor `BitWriter` class?

ShadauxCat replied:
It can't be renamed. The point is that it's called automatically via `using()`. As I explained in a previous comment, ref structs cannot implement interfaces, but c# has special handling to support `using` syntax for any red structs that implement a `Dispose()` method.

0xFA11 replied:
huh, interesting. OK :)

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs:
```diff
@@ -0,0 +1,312 @@
+    /// things to happen because the struct's lifetime could outlive the Reader/Writer's.)
+    /// </summary>
+    /// <typeparam name="TImplementation">The implementation struct</typeparam>
+    public ref struct BufferSerializer<TImplementation> where TImplementation : IBufferSerializerImplementation
```
I see a similar conversation initiated by Luke above but I'd like to kick-off another one.
Can we compare `if (IsReading)` approach vs interface then reader or writer impl behind the scenes approach to actually see performance difference?
If it's not that different, I'd suggest keeping things simple, less abstract and in one place (hence the existing `NetworkSerializer` implementation).

ShadauxCat replied:
Branching is pretty expensive. IIRC when I compared it before it was like a 2-3x cost increase to branch. I'll rerun that test in the morning since it's been a while.

ShadauxCat replied:
Responded to this in the above comment.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReaderExtensions.cs:
```diff
@@ -0,0 +1,286 @@
+                value = nb;
+                return;
+            }
+            /*if (value is INetworkSerializable)
```
👀

ShadauxCat replied:
This is to be implemented in IMessage work. I can't change the definition of INetworkSerializer to support FastBuffer* until all the existing uses of INetworkSerializable are gone.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReaderExtensions.cs:
```diff
@@ -0,0 +1,286 @@
+        /// for multiple reads at once by calling TryBeginRead.
+        /// </summary>
+        /// <param name="value">value to read</param>
+        public static void ReadValueSafe(this ref FastBufferReader reader, out GameObject value)
+        {
+            reader.ReadValueSafe(out ulong networkObjectId);
+
+            if (NetworkManager.Singleton.SpawnManager.SpawnedObjects.TryGetValue(networkObjectId, out NetworkObject networkObject))
+            {
+                value = networkObject.gameObject;
+                return;
+            }
```
we intentionally killed GameObject, NetworkObject & NetworkBehaviour serialization from built-in serializers. we wanted to kill NetworkManager/SpawnManager dependency in a primitive impl like this to keep it isolated and dependency-free (as much as we can).
also we were planning to offer more focused NetworkObject/NetworkBehaviour wrappers which would implement INetworkSerializable interface: `struct NetworkObjectReference : INetworkSerializable` etc.

ShadauxCat replied:
I agree with that approach... I'll look into doing that in the INetworkMessage branch when I refactor INetworkSerializable.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -0,0 +1,772 @@
+        /// <summary>
+        /// Frees the allocated buffer
+        /// </summary>
+        public unsafe void Dispose()
```
when and where you'd imagine this `Dispose()` method to be called?
I'm not seeing a destructor/finalizer calling this and wondering if this would mean we're pushing memory management onto users' plate?

ShadauxCat replied:
It'll be called by the messaging system whenever a buffer is no longer needed. I implemented it as IDisposable to support generic use of the serializer outside the messaging system since tools team has already requested the ability to do that for, i.e., file IO.

0xFA11 replied:
my worry is that we're not applying [idiomatic disposing pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) here.
we should call `Dispose()` from the destructor/finalizer and also implement `Dispose(bool disposing)` methods to call them delicately as `Dispose(false)` and `Dispose(true)` in the right contexts.

ShadauxCat replied:
Structs can't have finalizers/destructors. Only classes can have those.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/SerializationTypeTable.cs:
```diff
@@ -0,0 +1,449 @@
+    /// Serializers and Deserializers for FastBufferWriter and FasteBufferReader
+    /// SerializersPacked and DeserializersPacked for BytePacker and ByteUnpacker
+    /// </summary>
+    public static class SerializationTypeTable
```
so we're basically saying you can have completely separate serializers and deserializers for the same type? and even completely separate packed and non-packed serializers and deserializers? (in total 2 * 2 = 4 tables per type)

ShadauxCat replied:
Honestly, support for serializing `object` is only here because I don't know 100% for sure that I can kill it. I'll know when the IMessage stuff is done... And I'm hoping when I get there, I can just delete this file along with `ReadObject` and `WriteObject` entirely.

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Editor/Serialization/BitWriterTests.cs:
```diff
@@ -0,0 +1,332 @@
+                Assert.AreEqual(0b1010_10001011, *asInt);
+
+                writer.WriteByte(0b11111111);
+                Assert.AreEqual(0b11111111_00001010_10001011, *asInt);
```
oh wow! I didn't know C# had this type of cool binary literals <3

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/BytePacker.cs:
```diff
@@ -0,0 +1,617 @@
+        /// If true, an extra byte will be written to indicate whether or not the value is null.
+        /// Some types will always write this.
+        /// </param>
+        public static void WriteObjectPacked(ref FastBufferWriter writer, object value, bool isNullable = false)
```
since we're not offering anything automagical under this method here (apart from auto type -> method convenience), would it make more sense to drop this `WriteObject`/`ReadObject` combo entirely? I think we could have a convo about this and hopefully I might convince people to do so :)

ShadauxCat replied:
See above comment. I'm hoping the answer is yes but don't want to commit to that until I've converted the existing messages without needing it.

ShadauxCat committed:
- Removed RefArray
- Fixed incorrect text in a warning in FastBufferReader

ShadauxCat committed:
Removed RefArray meta file that stuck around.

ShadauxCat committed:
Review feedback: Used nameof() instead of string literal.

ShadauxCat committed:
-Removed BytewiseUtility.FastCopyBytes
-Added documentation on PreChecked() functions.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs:
```diff
@@ -0,0 +1,343 @@
+            m_Implementation.SerializeValue(ref value);
+        }
+
+        /// <summary>
+        /// Serialize a GameObject
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref GameObject value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
+
+        /// <summary>
+        /// Serialize a NetworkObject
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref NetworkObject value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
+
+        /// <summary>
+        /// Serialize a NetworkBehaviour
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref NetworkBehaviour value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
```
I wouldn't re-introduce these (both here and also reader/writer impls) because we know that we want to have `struct NetworkObjectReference` etc. for these scenarios. Killing these here and in reader/writer also kills NetworkManager.SpawnManager deps too — so, an easy win-win?

0xFA11 replied:
(I'd also add Read/Write Object to this context too, I remember calling that out earlier but I wanted to call out again to stress a little bit more)

ShadauxCat replied:
I'll say the same thing I said before.. I want to finish INetworkMessage implementation before I kill these. If it turns out they're unused/unneeded once everything's done, I'll happily kill them, but I don't want to kill them now and have to resurrect them later. Converting NetworkObject and NetworkBehaviour to be INetworkSerializable should be trivial... GameObject not so much, but we can also just not support game objects and demand that the player send us the NetworkObject.

MFatihMAR commented on com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs:
```diff
@@ -0,0 +1,343 @@
+        /// operations in release builds.
+        ///
+        /// To get the correct size to check for, use FastBufferWriter.GetWriteSize(value) or
+        /// FastBufferWriter.GetWriteSize&lt;type&gt;()
```
super-nit:
```suggestion
        /// FastBufferWriter.GetWriteSize<Type>()
```

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Editor/com.unity.netcode.editortests.asmdef:
```diff
@@ -3,12 +3,23 @@
-    ]
+    ],
+    "excludePlatforms": [],
+    "allowUnsafeCode": true,
```
nit: all we need is `"allowUnsafeCode": true` line manually added here (I don't like Unity auto-gen'ing these things all the time, leaving minimal stuff here makes us more future-proof between different Unity Editor versions)

0xFA11 reviewed (approved):
(ship-it-parrot)

ShadauxCat committed:
removed .gitignore.

ShadauxCat committed:
Merge branch 'develop' into feature/fast_buffer_reader_writer

# Conflicts:
#	com.unity.netcode.gameobjects/Tests/Editor/com.unity.netcode.editortests.asmdef

ShadauxCat committed:
Fixed compile errors that somehow didn't happen on my machine until I looked at the files they were in?

ShadauxCat committed:
standards.py --fix ... again ... despite no changes since the last success other than integrating develop ...

ShadauxCat committed:
Merge branch 'develop' into feature/fast_buffer_reader_writer
